game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java

```diff
@@ -80,7 +81,10 @@ public void draw(final Rectangle bounds, final GameData data, final Graphics2D g
       x = namePlace.get().x;
       y = namePlace.get().y;
     } else {
-      final Rectangle territoryBounds = getBestTerritoryNameRect(mapData, territory, fm);
+      if (territoryBounds == null) {
+        // Cache the bounds since re-computing it is expensive.
+        territoryBounds = getBestTerritoryNameRect(mapData, territory, fm);
+      }
```

Comment:
Via tracing, it appears that `draw()` can be called by multiple threads.  In that case, you might want to consider some form of synchronization when reading/writing `territoryBounds`.  Without it, there's the possibility that the cached reference won't be published to other threads.  I don't think that would affect correctness; it just means the rectangle may be calculated more than once.  Not sure if that's a big deal.

Reply:
FWIW, here's a possible solution that (I think) is thread-safe:

```diff
diff --git a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java
index a632f877f..642ce0ba6 100644
--- a/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java
+++ b/game-core/src/main/java/games/strategy/triplea/ui/screen/drawable/TerritoryNameDrawable.java
@@ -8,6 +8,10 @@ import java.awt.Polygon;
 import java.awt.Rectangle;
 import java.util.List;
 import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+import com.google.common.base.Suppliers;
 
 import games.strategy.engine.data.GameData;
 import games.strategy.engine.data.PlayerID;
@@ -21,7 +25,7 @@ import games.strategy.triplea.ui.mapdata.MapData;
 public class TerritoryNameDrawable implements IDrawable {
   private final String territoryName;
   private final UiContext uiContext;
-  private Rectangle territoryBounds;
+  private final AtomicReference<Supplier<Rectangle>> territoryBoundsSupplier = new AtomicReference<>();
 
   public TerritoryNameDrawable(final String territoryName, final UiContext uiContext) {
     this.territoryName = territoryName;
@@ -81,10 +85,8 @@ public class TerritoryNameDrawable implements IDrawable {
       x = namePlace.get().x;
       y = namePlace.get().y;
     } else {
-      if (territoryBounds == null) {
-        // Cache the bounds since re-computing it is expensive.
-        territoryBounds = getBestTerritoryNameRect(mapData, territory, fm);
-      }
+      territoryBoundsSupplier.compareAndSet(null, Suppliers.memoize(() -> getBestTerritoryNameRect(mapData, territory, fm)));
+      final Rectangle territoryBounds = territoryBoundsSupplier.get().get();
       x = territoryBounds.x + (int) territoryBounds.getWidth() / 2 - fm.stringWidth(territory.getName()) / 2;
       y = territoryBounds.y + (int) territoryBounds.getHeight() / 2 + fm.getAscent() / 2;
     }
```

Although, it's a bit ugly due to the initializer depending on local state.

(Note that the `Supplier` returned by `Suppliers#memoize()` is thread-safe with respect to the cached value.)

Reply:
Actually, no synchronization may be necessary after all.  Looking up the call stack from `draw()`, it appears the `GameData` lock is being acquired while drawing the image:

https://github.com/triplea-game/triplea/blob/2ff3dd44355d5816287c1bb81a3af44f3cd04ee6/game-core/src/main/java/games/strategy/triplea/ui/MapPanel.java#L575-L583

That introduces a memory barrier and should force the cached value to be published to any other pool thread that happens to access the object in the future.

**TL;DR:** Nevermind. :smile: